File: mergeSort.js
function mergeSort(array) {
  if (array.length < 2) return array;
  let mid = Math.floor(array.length / 2);
  let leftArray = array.slice(0, mid);
  let rightArray = array.slice(mid, array.length);
  return merge(mergeSort(leftArray), mergeSort(rightArray));
}
function merge(leftArray, rightArray) {
  let ans = [];
  while (leftArray.length && rightArray.length) {
    if (leftArray[0] < rightArray[0]) {
      ans.push(leftArray.shift());
    } else {
      ans.push(rightArray.shift());
    }
  }
  return [...ans, ...leftArray, ...rightArray];
}
let array = [
  1, 2, 3, 34, 5, 56, 7, 4, 0, 343, 2, 21, 1, 1, 23, 34, 9, -3, 986, 7, 53,
];
console.log(mergeSort(array));

File: mergeSort.cpp
#include<bits/stdc++.h>
using namespace std;
vector<long long >merge(vector<long long >leftArray, vector<long long>rightArray){
    vector<long long>ans;
    long long i=0,j=0;
    while(i<leftArray.size() && j<rightArray.size()){
        if(leftArray[i]<rightArray[j]){
            ans.push_back(leftArray[i++]);
        }
        else{
            ans.push_back(rightArray[j++]);
        }
    }
    while(i<leftArray.size()){
        ans.push_back(leftArray[i++]);
    }
    while(j<rightArray.size()){
        ans.push_back(rightArray[j++]);
    }
    return ans;
}
vector<long long>mergeSort(vector<long long>& array){
    if(array.size()<2)return array;
    vector<long long>leftArray(array.size()/2);
    vector<long long>rightArray(array.size()-leftArray.size());
    for(long long i=0;i<array.size();i++){
        if(i<leftArray.size()){
            leftArray[i]=array[i];
        }
        else{
            rightArray[i-leftArray.size()]=array[i];
        }
    }
    return merge(mergeSort(leftArray),mergeSort(rightArray));
}
int main(){
    vector< long long> array={1,2,3,4,5,6,7,3,2,2,4,4,5,32,2,23,4,423,4,8,4,0};
    for(auto&it: array)cout<<it<<' ';
    cout<<endl;
    for(auto&it:mergeSort(array))cout<<it<<' ';
}